### PR TITLE
[JENKINS-27395] Record flow node and enclosing stage/branch for tests

### PR DESCRIPTION
**Pull request by abayer**
[JENKINS-27395](https://issues.jenkins-ci.org/browse/JENKINS-27395)

Downstream of https://github.com/jenkinsci/junit-plugin/pull/76. This
records the `FlowNode#id` for the step that's actually recording the
test results and the enclosing stage(s)/branch(es) of that `FlowNode`
with each suite, along with the names of the enclosing
stage(s)/branch(es) on each `CaseResult` for display purposes.

This is accomplished by getting the `FlowNode` via
`context.get(FlowNode.class)`, and then passing the configured
`JUnitResultArchiver`, the node's ID, the enclosing stage/branch IDs,
and the run (which is used to get the `Run#getExternalizableId()` for
internal purposes) to the new `JUnitResultArchiver#parseAndAttach`
static method. That method does the rest of the work.

Once this is all released, the classic test results pages will display
the test name with the enclosing stage/branch names prepended (if
there are test results from multiple stages/branches in the full test
result - i.e., only if relevant). Blue Ocean has a PR to do the same
that's pending.

cc @reviewbybees esp @cyrille-leclerc @svanoort

### CODE REVIEW

**Comment by alvarolobato** [comment]
Let's make sure all these dependency bumps are in the envelope before merging and releasing.

**Comment by abayer** [comment]
@cyrille-leclerc rebased on master, bumped `junit` to 1.23 and updated API calls accordingly. Assuming tests pass, should be good to go.

File: jenkins-plugin/pom.xml [hunk 1]
```diff
@@ -77,6 +77,7 @@
     <scm-api-plugin.version>2.1.1</scm-api-plugin.version>
     <git-plugin.version>3.3.0</git-plugin.version>
     <junit-plugin.version>1.23</junit-plugin.version>
+    <java.level>7</java.level>
```
Comment (cyrille-leclerc) [inline jenkins-plugin/pom.xml]: I thought that Jenkins required Java 8. Am I wrong?

File: jenkins-plugin/pom.xml [hunk 2]
```diff
@@ -320,6 +335,157 @@
         <artifactId>ssh-credentials</artifactId>
         <version>1.13</version>
       </dependency>
+      <dependency>
```
Comment (cyrille-leclerc) [inline jenkins-plugin/pom.xml]: @abayer can you please tell us why you need to add 20+ dependencies including things that seem to be completely unrelated to the purpose of this plugin such as `commons-fileupload` or `ant`

File: jenkins-plugin/pom.xml (hunk 1 above)
Comment (abayer) [reply jenkins-plugin/pom.xml]: Starting with 2.60, yes, but since this uses core 2.46.3, it should still be Java 7.

File: jenkins-plugin/pom.xml (hunk 2 above)
Comment (abayer) [reply jenkins-plugin/pom.xml]: First, I added them to `<dependencyManagement>` - i.e., it just says "if a dependency of ours depends on this artifact, use this version". Second, it was needed to get the upper bounds enforcer checks that were added to the plugin parent to pass. @jglick can explain the rationale for that change, but the gist is that you already depended on all these artifacts (and many, many, many more - 339 total by my count!) via transitive dependencies, mainly from `maven-plugin` and `maven-invoker-plugin` as well as core, but the versions were inconsistent and needed to be rationalized.

Comment (cyrille-leclerc) [reply jenkins-plugin/pom.xml]: I looked too quickly at the diff, seem to be in the `<dependencyManagement>` section, not in the `<dependencies>`section. I need to double check

Comment (cyrille-leclerc) [reply jenkins-plugin/pom.xml]: Thanks for the explanation @abayer, sorry for the time you spent in this explanation, I wrote my comment "I need to double check" at the same time you were writing your explanation.

**Comment by cyrille-leclerc** [comment]
PR not forgotten, we want first to release 3.2.0 which is "alpha" for the moment.
